lightdash/lightdash
# fix: replace name of x and y fields on tc change

Relates to: #7996 

### Description:

Most of the code here comes from: https://github.com/lightdash/lightdash/pull/9269 - which was reverted - with the following fix:

- Correctly checks if the `chartConfigs` have changed to then update either the `useCartesianChartConfig` hook state or the `prop` via `VisualizationConfigCartesian` component.
    - Updates hook state when the valid config hasn't changed but the prop has
    - Updates prop when it hasn't changed but the hook changed the config

Steps to reproduce:
1. Apply this diff to the repo:
```diff
diff --git a/packages/e2e/cypress/e2e/app/createProjects.cy.ts b/packages/e2e/cypress/e2e/app/createProjects.cy.ts
index 463e86c2d..ecf002ab3 100644
--- a/packages/e2e/cypress/e2e/app/createProjects.cy.ts
+++ b/packages/e2e/cypress/e2e/app/createProjects.cy.ts
@@ -463,7 +463,7 @@ describe('Create projects', () => {
             testCustomDimensions(projectUuid);
         });
     });
-    it('Should create a Redshift project', () => {
+    it.only('Should create a Redshift project', () => {
         // https://docs.aws.amazon.com/redshift/latest/dg/c_redshift-and-postgres-sql.html
         // Amazon Redshift is based on PostgreSQL
         // So we can use our own PostgreSQL local instance to test the connection against Redshift
@@ -478,9 +478,9 @@ describe('Create projects', () => {
         configurePostgresWarehouse(warehouseConfig.redshift);

         testCompile().then((projectUuid) => {
-            testFilterStringEscaping(projectUuid);
-            testTimeIntervalsResults(projectUuid);
-            testPercentile(projectUuid);
+            // testFilterStringEscaping(projectUuid);
+            // testTimeIntervalsResults(projectUuid);
+            // testPercentile(projectUuid);

             createCustomDimensionChart(projectUuid);
             testCustomDimensions(projectUuid);
```
2. Run `yarn e2e-open`
3. Run `createProjects.cy.ts` e2e test

Before:
![image](https://github.com/lightdash/lightdash/assets/22939015/2ae7b478-fcb6-40bb-8fc3-f9ebcf441e2d)

After:
![image](https://github.com/lightdash/lightdash/assets/22939015/72f9821b-fd47-4fba-89aa-50da983b83f4)

### Reviewer actions
